### PR TITLE
PR for adding retry and warmup logic to application

### PR DESCRIPTION
**Pull request by Ajay-MS**

### CODE REVIEW

File: Tasks/Common/webdeployment-common/deployusingmsdeploy.ts [hunk 1]
```diff
@@ -65,4 +72,25 @@ export async function DeployUsingMSDeploy(webDeployPkg, webAppName, publishingPr
...
-}
\ No newline at end of file
+}
+
+export async function executeMSDeploy(msDeployCmdArgs, retryCount, retryInterval, errObj) {
+    var deferred = Q.defer();
+    try {
+        await tl.exec("msdeploy", msDeployCmdArgs, <any>{failOnStdErr: true, errStream: errObj});
+        deferred.resolve("Azure App service successfully deployed");
+    } catch (error) {
```
Comment (vincentdass) [inline Tasks/Common/webdeployment-common/deployusingmsdeploy.ts]: Can we invoke the retry logic, only when it is related to connection terminated.
We can also [inform the user](https://github.com/Microsoft/vsts-tasks/blob/2b77f875ea4ba5f78edd740c3ac28c60c730fcd2/Tasks/Common/webdeployment-common/msdeployutility.ts#L169) that we retry for the above error.

File: Tasks/Common/webdeployment-common/deployusingmsdeploy.ts [hunk 2]
```diff
@@ -65,4 +72,25 @@ export async function DeployUsingMSDeploy(webDeployPkg, webAppName, publishingPr
...
+        if (retryCount-- > 0) {
+             setTimeout(async function() {
+                 try {
+                    await executeMSDeploy(msDeployCmdArgs, retryCount, retryInterval, errObj);
+                 } catch (error) {
+                    deferred.reject(error);
+                 }
+             }, retryInterval);
```
Comment (vincentdass) [inline Tasks/Common/webdeployment-common/deployusingmsdeploy.ts]: Is retryInterval for 1 sec ?

File: Tasks/Common/webdeployment-common/deployusingmsdeploy.ts [hunk 3]
```diff
@@ -29,33 +33,36 @@ export async function DeployUsingMSDeploy(webDeployPkg, webAppName, publishingPr
...
+
     errObj.on('finish', () => {
         msDeployUtility.redirectMSDeployErrorToConsole();
     });
 
+    var retryCountParam = tl.getVariable("appservice.msdeployretrycount");
+    var retryCount = (retryCountParam && !(isNaN(Number(retryCountParam)))) ? Number(retryCountParam): DEFAULT_RETRY_COUNT; 
+
     try {
-        await tl.exec("msdeploy", msDeployCmdArgs, <any>{failOnStdErr: true, errStream: errObj});
```
Comment (vincentdass) [inline Tasks/Common/webdeployment-common/deployusingmsdeploy.ts]: Is '`<any>{failOnStdErr: true, errStream: errObj}`' not required ?

Comment (Ajay-MS) [inline Tasks/Common/webdeployment-common/deployusingmsdeploy.ts]: It has not been removed. It moved to inner function.

File: Tasks/Common/webdeployment-common/deployusingmsdeploy.ts (hunk 2 above)
Comment (Ajay-MS) [inline Tasks/Common/webdeployment-common/deployusingmsdeploy.ts]: As discussed will increase retry interval

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts [hunk 4]
```diff
@@ -123,6 +123,16 @@ async function run() {
             tl.setVariable(webAppUri, publishingProfile.destinationAppUrl);
         }
 
+        if(publishingProfile && publishingProfile.destinationAppUrl) {
+            try{
+                await azureRESTUtility.warmupAzureAppService(publishingProfile.destinationAppUrl);
+                console.log(tl.loc("appServiceWarmupedUpSuccessfully"));
+            } catch (error) {
+                console.log(tl.loc('failedToWarmupAppService'));
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: Do we need this in console. Warm is just as best case to increase the success rate. Isn't it enough to be in debug logs alone.

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts [hunk 5]
```diff
@@ -123,6 +123,16 @@ async function run() {
             tl.setVariable(webAppUri, publishingProfile.destinationAppUrl);
         }
 
+        if(publishingProfile && publishingProfile.destinationAppUrl) {
+            try{
+                await azureRESTUtility.warmupAzureAppService(publishingProfile.destinationAppUrl);
+                console.log(tl.loc("appServiceWarmupedUpSuccessfully"));
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: Do we need this console output?

File: Tasks/AzureRmWebAppDeployment/task.json [hunk 6]
```diff
@@ -375,6 +372,8 @@
         "ScriptStatusTimeout": "Unable to fetch script status due to timeout.",
         "PollingForFileTimeOut": "Unable to fetch script status due to timeout. You can increase the timeout limit by setting 'appservicedeploy.retrytimeout' variable to number of minutes required.",
         "InvalidPollOption": "Invalid polling option provided: %s.",
-        "MissingWebConfigParameters": "Some web.config parameters are missing"
+        "MissingWebConfigParameters": "Some web.config parameters are missing",
+        "appServiceWarmupedUpSuccessfully": "Successfully warmed up azure app service.",
```
Comment (kmkumaran) [inline Tasks/AzureRmWebAppDeployment/task.json]: I think the warm messages are only debug messages. The are not user facing and need not be localized.

File: Tasks/Common/azurerest-common/azurerestutility.ts [hunk 7]
```diff
@@ -491,3 +492,23 @@ export async function restartAppService(endpoint, resourceGroupName: string, web
...
+
+export async function warmupAzureAppService(webAppUrl) {
+    var deferred = Q.defer();
+    var headers = {};
+    httpObj.get('GET', webAppUrl, headers, (error, response, body) => {
+        if (error) {
+            tl.debug("Failed to warm up azure app service, error : " + error);
+            deferred.reject(error);
+        } else {
+            if(response.statusCode === 200) {
```
Comment (kmkumaran) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: Can we check from the response if the status of website is "running"?

File: Tasks/Common/webdeployment-common/deployusingmsdeploy.ts (hunk 1 above)
Comment (kmkumaran) [inline Tasks/Common/webdeployment-common/deployusingmsdeploy.ts]: ++. We Should retry only for the connection terminated error.

Comment (kmkumaran) [inline Tasks/Common/webdeployment-common/deployusingmsdeploy.ts]: Question - Is this specific to deployment using msdeploy? Are we not seeing this for Kudu based deployments?

Comment (kmkumaran) [inline Tasks/Common/webdeployment-common/deployusingmsdeploy.ts]: Can you add UT with mocks to simulate failures. Check with @vincentdass , he is writing a similar UT for mocking HTTP calls.

File: Tasks/Common/azurerest-common/azurerestutility.ts [hunk 8]
```diff
@@ -491,3 +492,23 @@ export async function restartAppService(endpoint, resourceGroupName: string, web
     });
     return deferred.promise;
 }
+
+export async function warmupAzureAppService(webAppUrl) {
+    var deferred = Q.defer();
+    var headers = {};
+    httpObj.get('GET', webAppUrl, headers,async (error, response, body) => {
```
Comment (vincentdass) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: is **async** required here ?

File: Tasks/Common/webdeployment-common/deployusingmsdeploy.ts [hunk 9]
```diff
@@ -29,40 +33,70 @@ export async function DeployUsingMSDeploy(webDeployPkg, webAppName, publishingPr
...
+    var errorFile = path.join(tl.getVariable('System.DefaultWorkingDirectory'),"error.txt");
+    var fd = fs.openSync(errorFile, "w");
+    var errObj = fs.createWriteStream("", {fd: fd} );
+
+    errObj.on('finish', async () => {
+        if(msDeployError) {
+            var shouldRetry = msDeployUtility.shouldRetryMSDeploy();
+            if(shouldRetry && retryCount-- > 0) {
+                try {
+                    await executeMSDeploy(msDeployCmdArgs, retryCount, retryInterval);
```
Comment (vincentdass) [inline Tasks/Common/webdeployment-common/deployusingmsdeploy.ts]: just a doubt.
Should we require defer.resolve(), after this line ?

File: Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts [hunk 10]
```diff
@@ -123,6 +123,15 @@ async function run() {
             tl.setVariable(webAppUri, publishingProfile.destinationAppUrl);
         }
 
+        if(publishingProfile && publishingProfile.destinationAppUrl) {
+            try{
+                await azureRESTUtility.warmupAzureAppService(publishingProfile.destinationAppUrl);
+                deployUtility.sleep(3000);
```
Comment (vincentdass) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: why sleep is required ?

Comment (Ajay-MS) [inline Tasks/AzureRmWebAppDeployment/azurermwebappdeployment.ts]: After warming up waiting for small period of time to proceed with deployment of application. Will discuss offline over it.

File: Tasks/Common/azurerest-common/azurerestutility.ts (hunk 8 above)
Comment (Ajay-MS) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: No required, will remove this.

File: Tasks/Common/webdeployment-common/msdeployutility.ts [hunk 11]
```diff
@@ -186,4 +186,22 @@ export function redirectMSDeployErrorToConsole() {
 
         tl.rmRF(msDeployErrorFilePath);
     }
+}
+
+export function shouldRetryMSDeploy() {
+    var msDeployErrorFilePath = tl.getVariable('System.DefaultWorkingDirectory') + '\\error.txt';
```
Comment (kmkumaran) [inline Tasks/Common/webdeployment-common/msdeployutility.ts]: make file name const.

File: Tasks/Common/azurerest-common/azurerestutility.ts [hunk 12]
```diff
@@ -491,3 +492,23 @@ export async function restartAppService(endpoint, resourceGroupName: string, web
     });
     return deferred.promise;
 }
+
+export async function warmupAzureAppService(webAppUrl) {
```
Comment (kmkumaran) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: Can you call this testAzureWebAppAvailability

File: Tasks/Common/azurerest-common/azurerestutility.ts [hunk 13]
```diff
@@ -13,6 +13,7 @@ var restObj = new restClient.RestCallbackClient(httpObj);
 
 var defaultAuthUrl = 'https://login.windows.net/';
 var azureApiVersion = 'api-version=2016-08-01';
+var defaultWebAppAvailabilityTimeout = 3000;
```
Comment (kmkumaran) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: Can you include the units in the name of variable for readability... defaultWebAppAvailabilityTimeoutInSecs or defaultWebAppAvailabilityTimeoutInMS

**Pull request merged**